### DISCUSSION TITLE
What does bufferedWrite do?

### DISCUSSION THREAD

**Discussion by gera2ld** [gildas-lormeau/zip.js#249] [Q&A] [answered]
Could you please explain what `bufferedWrite` exactly does?
I tried `writer.add` with/without `bufferedWrite` but it seems to make no difference, either in output or performance. I'm also wondering if it's safe to add multiple files in parallel with the default option `bufferedWrite: false`.

**Comment by gildas-lormeau** [comment, marked as answer]
When `bufferedWrite` is set to `true`, zip.js creates a separate buffer for each entry being created to avoid data corruption in the zip file. However, when it's set to `false`, zip.js can still detect if `ZipWriter#add` is called in parallel and create automatically the buffers if necessary. In this case, the entry added via the first call to `ZipWriter#add` is blocking because zip.js did not know there would be concurrent writes. So the performance depends on the size of the first entry added in the zip. That's the reason why you should set `bufferedWrite` to `true` when you know you will call `ZipWriter#add` multiple times in parallel. Any value is safe though.
Finally, note that the option `keepOrder` is also set to `true` by default. It's necessary when writing files that could rely on zip64. Since entries are ordered, this can have a similar impact on performance. If you know that the zip file does not need zip64 (e.g. if it weights less than 4GB), then you can set `keepOrder` to `false`. Thus, there won't be any blocking lock when adding entries in parallel.

**Comment by gera2ld** [comment]
Thanks a lot for the detailed clarification, that's very clear. I guess it would be great to have this in the documentation.

**Comment by gildas-lormeau** [comment]
I agree this should be documented. Meanwhile, I created a discussion thread here: https://github.com/gildas-lormeau/zip.js/discussions/237 in order to keep it visible.
